### ISSUE TITLE
Event Details - Workflows fails to show workflow information for scheduled events.

### ISSUE THREAD

**Issue by kristofkeppens** [opencast/opencast#1069]
**Describe the bug**
When going to event details ( trying this with failed events ) , workflows you should get an overview of the workflows and possible errors with the workflows. This works as intented for uploaded mediafiles but for scheduled events you get the workflow configuration view on that page, instead of the applied workflows and their status.

**Environment Information**
 - MacOS
 - Chrome
 - Opencast 7.2

**Comment by karendolan** [comment]
@flyapen I see this a lot in this and other similar situations. I suspect it is related to the varied and probabilistic ways of determining if an event is in a scheduled state, right ~90% of the time. Do you have a fix in mind? The following is what we are trying to get the probability up to 97%. And I see the missing 3% at the point of scheduling a new event and the event being fully owned by the scheduler service.

```   
public Source getEventSource(Event event) {
-    if (event.getWorkflowId() != null && isWorkflowActive(event.getWorkflowState()))
-      return Source.WORKFLOW;
 
-    if (event.getSchedulingStatus() != null && !event.hasRecordingStarted())
+    // #HUDCE OPC-298 "Admin UI not showing workflow when ingested event start date changed to future"
+    //
+    // Priority of Source Likelihood (i.e. enhanced #2 with an extra check on processed state)
+    //
+    // 1. If there is a running workflow, event source is the workflow service
+    // 2. If this is scheduled event and the start time is in the future ALSO check if it's actually already processed.
+    // to cover the case where the ingested pub is giving a future date (i.e. pre-recorded lectures). If the pub is not
+    // in processed state,then source is still the scheduler service.
+    // 3. If it's not in scheduler state and no running workflow, check if it's been archived, therefore archive source.
+    // 4. If not archived and not scheduled and not running workflow, but has a workflow Id, the workflow service knows
+    // it.
+    // 5. Last fallback, when no workflow, no archive, and no scheduled state, is that it's in the process of
+    // being scheduled.
+    logger.debug("Event '{}' archive '{}' latest workflow '{}' in state '{}', scheduling status '{}' event status '{}'",
+            event.getIdentifier(), event.getArchiveVersion(), event.getWorkflowId(), event.getWorkflowState(),
+            event.getSchedulingStatus(), event.getEventStatus());
+
+     if (event.getWorkflowId() != null && isWorkflowActive(event.getWorkflowState()))
+       return Source.WORKFLOW;
+
+    // #DCE OPC-298 Event's "processed" state overrules future start date.
+    if (event.getSchedulingStatus() != null && !event.hasRecordingStarted()
+            && ! Event.postIngestEventStatusList.contains(event.getEventStatus())) {
       return Source.SCHEDULE;
+    }
 
     if (event.getArchiveVersion() != null)
       return Source.ARCHIVE;
....
+++ b/modules/index-service/src/main/java/org/opencastproject/index/service/impl/index/event/Event.java
@@ -94,6 +94,21 @@ public class Event implements IndexObject {
...
+  /** #HUDCE OPC-298 post ingest event status */
+  public static final List<String> postIngestEventStatusList = new ArrayList<String>() {
+    {
+       add("EVENTS.EVENTS.STATUS.PENDING");
+       add("EVENTS.EVENTS.STATUS.PROCESSING");
+       add("EVENTS.EVENTS.STATUS.PAUSED");
+       add("EVENTS.EVENTS.STATUS.PROCESSED");
+       add("EVENTS.EVENTS.STATUS.PROCESSING_FAILURE");
+       add("EVENTS.EVENTS.STATUS.PROCESSING_CANCELED");
+       // #HUDCE specific states
+       add("EVENTS.EVENTS.STATUS.NEEDS_CUTTING");
+       add("EVENTS.EVENTS.STATUS.NEEDS_TEST_LOCAL");
+    }
+   };

...
   public boolean hasRecordingStarted() {
-    if (getSchedulingStatus() == null)
+    // #HUDCE OPC-298 add additional check for event already processed to know the recordings was already
+    // started at some past date.
+    if (getSchedulingStatus() == null || postIngestEventStatusList.contains(getEventStatus()))
       return true;
```

**Comment by staubesv** [comment]
As I've recently changed some methods that might be involved in causing the bug, I felt guilty enough to provide some analysis for the bug:

1. The variable `workflows.scheduling` determines whether you see the workflow list (https://github.com/opencast/opencast/blob/develop/modules/admin-ui/src/main/webapp/scripts/shared/partials/modals/event-details.html#L379) or the workflow configuration panel (https://github.com/opencast/opencast/blob/develop/modules/admin-ui/src/main/webapp/scripts/shared/partials/modals/event-details.html#L445)  in Events->Event Details->Workflows

2. The variable  `workflows.scheduling` is set here: https://github.com/opencast/opencast/blob/develop/modules/admin-ui/src/main/webapp/scripts/shared/resources/eventWorkflowsResource.js#L25

3. The Admin UI facade uses a custom logic to determine whether the event is still being scheduled or not:
https://github.com/opencast/opencast/blob/develop/modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java#L1555

4. You find the methods used in step 3 here: https://github.com/opencast/opencast/blob/develop/modules/index-service/src/main/java/org/opencastproject/index/service/impl/index/event/Event.java#L906 and https://github.com/opencast/opencast/blob/develop/modules/index-service/src/main/java/org/opencastproject/index/service/impl/index/event/Event.java#L917

The cause for the symptom described by this issue is that the Admin UI facade uses a custom logic in step 3 to determine whether the event is still "Scheduled"  vs. the actual event status as displayed in the events table.
That way, the event status might be displayed as "Processing failure" in the events table, but looks like "Scheduled" for Events->Event Details->Workflows, resulting in the wrong page being displayed.

Without thinking too much about it, it seems tempting to directly use the actual event status (https://github.com/opencast/opencast/blob/develop/modules/index-service/src/main/java/org/opencastproject/index/service/impl/index/event/Event.java#L946)  in step 3 instead of a custom method to determine another variant of the event status as currently done. 

That way, you would always (100.0%) get the correct page in Events->Event Details->Workflows for a given actual event status.

Whether the actual event status displayed in the events table is determined correctly in 100% of the cases is another question, but would also be another issue ;-)

@flyapen The proposed solution above would work for your case as only events with the event status _Scheduling_ would show the workflow configuration panel in Events->Event Details->Workflows (with 100% accuracy)

@karendolan Do you agree?

**Comment by kristofkeppens** [comment]
@staubesv I've changed https://github.com/opencast/opencast/blob/develop/modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java#L1555 to directly use the event status and this works as expected ( in our case ). I can create a pull request if this is a solution that works for most cases, I don't know all the reason's why there is a separate logic used there so I might have broken something ( at least nothing visible in our installation ).

**Comment by staubesv** [comment]
@flyapen I don't think that there is any good reason to use the separate logic and I am sure it works for all cases. Happy to hear you implemented a solution - please do raise a pull request.

**Comment by gregorydlogan** [comment]
Fixed by `0a7e26ecdf8d446aaf148d8c17d0457284633d37`, merged as part of #1122
